### ISSUE TITLE
Details around Platt scaling and open-sourcing zero-shot transfer

### ISSUE THREAD

**Issue by sayakpaul** [facebookresearch/SWAG#13]
Thank you for open-sourcing this work. 

From the paper:

![image](https://user-images.githubusercontent.com/22957388/209773816-fc85bd14-aeb0-4a59-b58e-a15144397918.png)

* What does $C$ contain? The classes on which SWAG pre-training was performed? 
* How is $p$ obtained? 

Is it possible to also accompany this repository with a notebook that shows how to adapt the pre-trained models for zero-shot transfer? I don't think there is a good example available on the internet that shows how to extend an image recognition model like SWAG for zero-shot classification. 

@mannatsingh

**Comment by mannatsingh** [comment]
@sayakpaul absolutely, glad you find the work useful!

* You are right, $C$ contains the classes (hashtags) on which SWAG pre-training was performed - there were about 27K classes
* $p$ is just the output probability distribution from the final linear layer corresponding to the $C$ classes (followed by softmax), and $p'$ is a uniform distribution.

Overall, the Platt Scaling "fixes" some of the imbalance in the hashtag distribution by producing more uniform outputs. After this, to map the results to IN1K classes we use WordNet similarities. Unfortunately as mentioned in the broader impact section of the paper, we did not release the final hashtag prediction layer, so we won't be able to produce a notebook for zero-shot classification. 

I do think that there might be other ways to utilize a broad output vocabulary (such as SWAG's) for "transferring" to downstream datasets (as long as there is concept overlap), but our exploration was limited.

**Comment by sayakpaul** [comment]
Thanks!

I am guessing here $p$ is computed on the test images (available during the zero-shot transfer)? 

Also, could you shed a bit more detail on how the similarity search was performed with WordNet?

**Comment by mannatsingh** [comment]
Sorry for the delay in the response, I was on time off!

Yes, that is correct, $p$ is computed on the full test set. 

For the similarity, ImageNet is a special "easy" case since it maps directly to WordNet synsets (and so do our hashtags). So ultimately, we just need similarity between synsets to get a mapping from hashtags to classes, which can be computed using the WordNet path similarity (https://www.nltk.org/howto/wordnet.html#similarity). There's more details mentioned in Appendix D of the paper which primarily deal with handling the many-to-many nature of such a mapping.

**Comment by sayakpaul** [comment]
Thanks, Mannat. I appreciate the details very much. Closing this.
